ethereum/datastructures/src/main/java/tech/pegasys/teku/datastructures/util/LengthBoundCalculator.java

```diff
@@ -0,0 +1,157 @@
+        vectorCount++;
+
+      } else if (isBitvector(fieldType)) {
+        fieldLengthBounds = calculateBitvectorLength(reflectionInfo, bitvectorCount);
+        bitvectorCount++;
+
+      } else if (isPrimitive(fieldType)) {
+        fieldLengthBounds = new LengthBounds(getPrimitiveLength(fieldType));
+
+      } else {
+        throw new IllegalArgumentException(
+            "Don't know how to calculate length for " + fieldType.getSimpleName());
+      }
+
+      if (isVariable(fieldType)) {
+        variableFieldCount++;
+        // The fixed parts includes an offset in place of the variable length value
+        lengthBounds = lengthBounds.add(new LengthBounds(BYTES_PER_LENGTH_OFFSET.longValue()));
+      }
+      lengthBounds = lengthBounds.add(fieldLengthBounds);
+    }
+    return lengthBounds;
+  }
+
+  private static LengthBounds calculateBitvectorLength(
+      final ReflectionInformation reflectionInfo, final int bitvectorCount) {
+    final LengthBounds fieldLengthBounds;
+    final Integer size = reflectionInfo.getBitvectorSizes().get(bitvectorCount);
+    final int serializationLength = Bitvector.sszSerializationLength(size);
+    fieldLengthBounds = new LengthBounds(serializationLength, serializationLength);
+    return fieldLengthBounds;
+  }
+
+  private static LengthBounds calculateSszVectorLength(
+      final ReflectionInformation reflectionInfo, final int vectorCount) {
+    final LengthBounds fieldLengthBounds;
+    final Class<?> elementType = reflectionInfo.getVectorElementTypes().get(vectorCount);
+    final int vectorLength = reflectionInfo.getVectorLengths().get(vectorCount);
+    final LengthBounds elementLengthBounds = getElementLengthBounds(elementType);
+    fieldLengthBounds =
+        new LengthBounds(
+            vectorLength * elementLengthBounds.getMin(),
+            vectorLength * elementLengthBounds.getMax());
+    return fieldLengthBounds;
+  }
+
+  private static LengthBounds calculateSszListLength(
+      final ReflectionInformation reflectionInfo, final int variableFieldCount) {
+    final LengthBounds fieldLengthBounds;
+    final Class<?> listElementType = reflectionInfo.getListElementTypes().get(variableFieldCount);
+    final long listElementMaxSize = reflectionInfo.getListElementMaxSizes().get(variableFieldCount);
+    final LengthBounds elementLengthBounds = getElementLengthBounds(listElementType);
+    final long variableFieldOffsetsLength =
+        isVariable(listElementType) ? BYTES_PER_LENGTH_OFFSET.intValue() * listElementMaxSize : 0;
+    fieldLengthBounds =
+        new LengthBounds(
+            0, elementLengthBounds.getMax() * listElementMaxSize + variableFieldOffsetsLength);
+    return fieldLengthBounds;
+  }
+
+  private static LengthBounds calculateBitlistLength(
+      final ReflectionInformation reflectionInfo, final int variableFieldCount) {
+    final LengthBounds fieldLengthBounds;
+    final long maxSize = reflectionInfo.getBitlistElementMaxSizes().get(variableFieldCount);
```

Comment:
Not sure here: `getBitlistElementMaxSizes()` seems to return _only_ `Bitlist` instance sizes in the referred class. Why is it indexed by `variableFieldCount` which is a total counter of lists and bitfields?

Reply:
That is *highly* suspicious but exactly matches what `SimpleOffsetSerializer` does. It also works for every type we know how to SSZ serialise, but I strongly suspect that's a coincidence and there just aren't any objects that have both `SSZList` and `Bitlist`.

Created a test class with both and sure enough it fails with `IndexOutOfBoundsException`.  So have fixed both `LengthBoundsCalculator` and `SimpleOffsetSerializer`.